### ISSUE TITLE
BUG: Syncing takes forever.

### ISSUE THREAD

**Issue by stonelazy** [neptune-ai/neptune-client#808]
### Describe the bug
Am using Pytorch lightning for training. Once the training is complete, am getting the dialog   
`Still waiting for the remaining operations to complete` and this is going on forever ( atleast 2days now) 
### Reproduction
This isn't reproducible all the time, only under certain cases.. In the past 2 months i would have faced this 3-4 times.

### Traceback
<img width="893" alt="image" src="https://user-images.githubusercontent.com/11377834/148496061-89d8c9ed-abe6-42b9-8ee1-9fd7389ce7f7.png">
### Environment
Collecting environment information...
PyTorch version: 1.10.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.31

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.11.0-38-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: RTX A6000
GPU 1: RTX A6000
GPU 2: RTX A6000
GPU 3: RTX A6000
GPU 4: RTX A6000
GPU 5: RTX A6000
GPU 6: RTX A6000
GPU 7: RTX A6000

Nvidia driver version: 460.91.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy==0.910
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.1
[pip3] pytorch-lightning==1.5.7
[pip3] torch==1.10.0+cu111
[pip3] torch-poly-lr-decay==0.0.1
[pip3] torchaudio==0.10.0+cu111
[pip3] torchmetrics==0.6.2
[pip3] neptune-client==0.12.1

**Comment by Blaizzy** [comment]
Hey @stonelazy,

I suspect what is making your script get stuck at times is the connection loss between your machine and the Neptune server (could be a connectivity issue or something else) and by default, the neptune-client has a retry timeout of **1 h**, so basically, it will wait and attempt to reconnect with the Neptune server for 1h.

Good thing is that you can override the timeout environment variable [NEPTUNE_SYNC_BATCH_TIMEOUT_ENV](https://github.com/neptune-ai/neptune-client/blob/ad93b9f32d5c4c86ab7606e2bc88606b3ab6514c/neptune/new/sync.py#L58) from the default **1 h** (3600 s) to **1 min** (60 s) for example.

Example
`export NEPTUNE_SYNC_BATCH_TIMEOUT_ENV=60`

This change will cause the neptune-client to attempt to reconnect for 60 seconds only and stop if not successful. In such cases where it fails to reconnect the tracked metadata will be saved locally and then you can manually sync all the locally saved metadata with the Neptune server using `neptune sync` command in your terminal. 
Example:
`neptune sync`

`neptune sync` command: https://docs.neptune.ai/you-should-know/connection-modes#synchronizing-data-with-neptune-servers

**Note**: Before running the sync command, first navigate to the current working directory of your project where the `.neptune` folder was created (in other words, the directory where you ran your script).

**Comment by Blaizzy** [comment]
Do you have any parallel connections/calls limits in your network? If so, what is it?

**Comment by Blaizzy** [comment]
By the way, I also noticed you are using an old version of neptune-client v0.12.0, I want to inform you that we released v0.14.1 last week which you can upgrade to.

`pip install -U neptune-client`

**Comment by Blaizzy** [comment]
Additionally,

I also got to know that if you stdout lots of data (especially progress bars such as tqdm), Neptune might take a long time to post-process.

**Comment by Blaizzy** [comment]
Hey @stonelazy 

Any updates?

**Comment by stonelazy** [comment]
Sorry for the delay, I wasn't able to reproduce this issue apart from that experiment. Also, i don't stdout any data apart from my loss value in progress bars.

**Comment by Blaizzy** [comment]
So this issue only happened once, right? Now everything is working fine?


> i don't stdout any data apart from my loss value in progress bars.

I want to understand this part better, so I have a couple of questions:
- Do you use `tqdm` to display your loss value in progress bars?
- Can you send me an example code and stdout output for this part?

**Comment by stonelazy** [comment]
> So this issue only happened once, right? Now everything is working fine?  

This issue happens occasionally. Not every time.

> I want to understand this part better, so I have a couple of questions:  

No, we use Pytorch Lightning's `prog_bar=True` option for logging the values in progress bars.

`        self.log(
            "train_loss",
            loss,
            on_step=True,
            on_epoch=True,
            prog_bar=True,
            logger=True,
            sync_dist=True
        )
`

**Comment by Blaizzy** [comment]
I think I got the issue. 
### Issue
So you are using multiple GPUs. Because of that, you set the `sync_dist` argument to `True`
In the PTL docs they mention the following:

> The sync_dist option can also be used in logging calls during the step methods, but be aware that this can lead to significant communication overhead and slow down your training.

Docs: https://pytorch-lightning.readthedocs.io/en/stable/advanced/multi_gpu.html#synchronize-validation-and-test-logging

Perhaps that should be in bold as it is more of a warning.

### Possible solution
They also mention that alternatively you can perform some computation manually and log the reduced result on rank 0.

```
def test_epoch_end(self, outputs):
    mean = torch.mean(self.all_gather(outputs))

    # When logging only on rank 0, don't forget to add
    # ``rank_zero_only=True`` to avoid deadlocks on synchronization.
    if self.trainer.is_global_zero:
        self.log("my_reduced_metric", mean, rank_zero_only=True)
```


Please try it and let me know if it completely solves this problem!

**Comment by stonelazy** [comment]
Not having this issue any more. Hence closing this._
